### PR TITLE
Add ImageConfig to the serialization context for dynamic task

### PR DESCRIPTION
**Pull request by pingsutw**
## Tracking issue
NA

## Why are the changes needed?
Dynamic task tries to build the ImageSpec at runtime. 

## What changes were proposed in this pull request?

- Build the ImageSpec for the task entities that are inside the dynamic task
- Add ImageConfig to the serialization context for dynamic tasks at compile time.
- At runtime, the dynamic task will look up the image URL in the ImageConfig instead of rebuilding the image spec.

## How was this patch tested?
remote cluster

### Setup process
```python
from click.testing import CliRunner

from flytekit import ImageSpec, task, dynamic, workflow
from flytekit.clis.sdk_in_container import pyflyte


new_flytekit = "git+https://github.com/flyteorg/flytekit.git@28d4da3a49c98a9f8dc93fef0945e22ca3cd208a"

repro_img = ImageSpec(
    builder="envd",
    packages=["mypy", new_flytekit],
    apt_packages=["git"],
    registry="pingsutw",
    source_root="."
)

image_foo = repro_img.with_apt_packages(["curl"])
image_d1 = repro_img.with_apt_packages(["wget"])
image_bar = repro_img.with_apt_packages(["vim"])


# hello
@task(container_image=image_foo)
def foo():
    print("foo")


@task(container_image=image_bar)
def bar():
    print("bar")


@dynamic(container_image=image_d1)
def d1():
    bar()


@workflow
def wf():
    foo()
    d1()
```

### Screenshots
Screenshot 2024-06-02 at 7 39 43 PM

Also tested it on serverless

Screenshot 2024-06-02 at 7 55 44 PM

## Check all the applicable boxes 

- [x] I updated the documentation accordingly.
- [x] All new and existing tests passed.
- [x] All commits are signed-off.

## Related PRs
NA

## Docs link
NA

### CODE REVIEW

**Comment by pingsutw** [comment]
cc @kumare3 @EngHabu

File: flytekit/tools/translator.py [hunk 1]
```diff
@@ -176,6 +177,15 @@ def get_serializable_task(
     )
 
     if isinstance(entity, PythonFunctionTask) and entity.execution_mode == PythonFunctionTask.ExecutionBehavior.DYNAMIC:
+        for e in context_manager.FlyteEntities.entities:
+            if isinstance(e, PythonAutoContainerTask):
+                # 1. Build the ImageSpec for the entities that are inside the dynamic task,
```
Comment (EngHabu) [inline flytekit/tools/translator.py:182]: How does that work? how do we know the entities inside an dynamic task at serialization time?

Comment (pingsutw) [reply flytekit/tools/translator.py:182]: actually, it builds all the entities in the [context](https://github.com/flyteorg/flytekit/blob/9703959508706ca6b3f39e665b3e38fd724978ff/flytekit/core/context_manager.py#L950), not only in dynamic

Comment (EngHabu) [reply flytekit/tools/translator.py:182]: In theory a dynamic wf may reference tasks that don't exist in the static context... We will say you will have to statically declare the task or have it use the same image as the dynamic task. Right?

**Review by EngHabu — APPROVED** [review]

**Review by wild-endeavor — DISMISSED** [review]
deduped

**Review by wild-endeavor — APPROVED** [review]

**Pull request merged**